File: hexameter-core/src/main/java/org/codetome/hexameter/core/internal/impl/HexagonImpl.java
package org.codetome.hexameter.core.internal.impl;

import static java.lang.Math.cos;
import static java.lang.Math.sin;

import java.awt.Rectangle;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.api.Hexagon;
import org.codetome.hexameter.core.api.HexagonOrientation;
import org.codetome.hexameter.core.api.Point;
import org.codetome.hexameter.core.api.contract.HexagonDataStorage;
import org.codetome.hexameter.core.api.contract.SatelliteData;
import org.codetome.hexameter.core.backport.Optional;
import org.codetome.hexameter.core.internal.GridData;

/**
 * Default implementation of the {@link Hexagon} interface.
 */
@SuppressFBWarnings("EI_EXPOSE_REP")
public final class HexagonImpl<T extends SatelliteData> implements Hexagon<T> {
    private final CubeCoordinate coordinate;
    private final transient Point[] pointArray;
    private final transient double[] vertices;
    private final transient List<Point> pointList;
    private final transient Rectangle externalBoundingBox;
    private final transient Rectangle internalBoundingBox;
    private final transient GridData sharedData;
    private final transient HexagonDataStorage<T> hexagonDataStorage;

    /**
     * Creates a new {@link Hexagon} object from shared data and a coordinate.
     *
     * @param gridData           grid data
     * @param coordinate         coordinate
     * @param hexagonDataStorage data map
     **/
    HexagonImpl(final GridData gridData, final CubeCoordinate coordinate, final HexagonDataStorage<T> hexagonDataStorage) {
        this.sharedData = gridData;
        this.coordinate = coordinate;
        this.hexagonDataStorage = hexagonDataStorage;

        this.pointList = calculatePoints();
        this.pointArray = getPoints().toArray(new Point[0]);
        final int x1 = (int) pointArray[3].getCoordinateX();
        final int y1 = (int) pointArray[2].getCoordinateY();
        final int x2 = (int) pointArray[0].getCoordinateX();
        final int y2 = (int) pointArray[5].getCoordinateY();

        externalBoundingBox = new Rectangle(x1, y1, x2 - x1, y2 - y1);
        internalBoundingBox = new Rectangle((int) (getCenterX() - (1.25 * sharedData.getRadius() / 2)),
                (int) (getCenterY() - (1.25 * sharedData.getRadius() / 2)),
                (int) (1.25f * sharedData.getRadius()),
                (int) (1.25f * sharedData.getRadius()));

        this.vertices = new double[]{
                pointArray[0].getCoordinateX(), pointArray[0].getCoordinateY(),
                pointArray[1].getCoordinateX(), pointArray[1].getCoordinateY(),
                pointArray[2].getCoordinateX(), pointArray[2].getCoordinateY(),
                pointArray[3].getCoordinateX(), pointArray[3].getCoordinateY(),
                pointArray[4].getCoordinateX(), pointArray[4].getCoordinateY(),
                pointArray[5].getCoordinateX(), pointArray[5].getCoordinateY(),
        };
    }

    private List<Point> calculatePoints() {
        final List<Point> points = new ArrayList<>();
        for (int i = 0; i < 6; i++) {
            final double angle = 2 * Math.PI / 6 * (i + sharedData.getOrientation().getCoordinateOffset());
            final double x = getCenterX() + sharedData.getRadius() * cos(angle);
            final double y = getCenterY() + sharedData.getRadius() * sin(angle);
            points.add(Point.fromPosition(x, y));
        }
        return points;
    }

    @Override
    public String getId() {
        return coordinate.toAxialKey();
    }

    @Override
    public List<Point> getPoints() {
        return pointList;
    }

    @Override
    public List<Point> getPointList() {
        return getPoints();
    }

    @Override
    public Point[] getPointArray() {
        return pointArray;
    }

    @Override
    public double[] getVertices() {
        return vertices;
    }

    @Override
    public Rectangle getExternalBoundingBox() {
        return externalBoundingBox;
    }

    @Override
    public Rectangle getInternalBoundingBox() {
        return internalBoundingBox;
    }

    @Override
    public CubeCoordinate getCubeCoordinate() {
        return coordinate;
    }

    @Override
    public int getGridX() {
        return coordinate.getGridX();
    }

    @Override
    public int getGridY() {
        return coordinate.getGridY();
    }

    @Override
    public int getGridZ() {
        return coordinate.getGridZ();
    }

    @Override
    public double getCenterX() {
        if (HexagonOrientation.FLAT_TOP.equals(sharedData.getOrientation())) {
            return coordinate.getGridX() * sharedData.getHexagonWidth() + sharedData.getRadius();
        } else {
            return coordinate.getGridX() * sharedData.getHexagonWidth() + coordinate.getGridZ()
                    * sharedData.getHexagonWidth() / 2 + sharedData.getHexagonWidth() / 2;
        }
    }

    @Override
    public double getCenterY() {
        if (HexagonOrientation.FLAT_TOP.equals(sharedData.getOrientation())) {
            return coordinate.getGridZ() * sharedData.getHexagonHeight() + coordinate.getGridX()
                    * sharedData.getHexagonHeight() / 2 + sharedData.getHexagonHeight() / 2;
        } else {
            return coordinate.getGridZ() * sharedData.getHexagonHeight() + sharedData.getRadius();
        }
    }

    @Override
    public Optional<T> getSatelliteData() {
        return hexagonDataStorage.getSatelliteDataBy(getCubeCoordinate());
    }

    @Override
    public void setSatelliteData(final T satelliteData) {
        this.hexagonDataStorage.addCoordinate(getCubeCoordinate(), satelliteData);
    }

    @Override
    public void clearSatelliteData() {
        this.hexagonDataStorage.clearDataFor(getCubeCoordinate());
    }

    @Override
    public int hashCode() {
        return Objects.hash(coordinate);
    }

    @Override
    public boolean equals(final Object object) {
        if (this == object) {
            return true;
        }
        if (object == null || getClass() != object.getClass()) {
            return false;
        }
        final HexagonImpl hexagon = (HexagonImpl) object;
        return Objects.equals(coordinate, hexagon.coordinate);
    }
}

File: mixite2.core/src/commonMain/kotlin/org/hexworks/mixite2/core/internal/impl/HexagonImpl.kt
package org.hexworks.mixite2.core.internal.impl

import org.hexworks.cobalt.datatypes.Maybe
import org.hexworks.mixite2.core.api.*
import org.hexworks.mixite2.core.api.contract.HexagonDataStorage
import org.hexworks.mixite2.core.api.contract.SatelliteData
import org.hexworks.mixite2.core.internal.GridData
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * Default implementation of the [Hexagon] interface.
 */
class HexagonImpl<T : SatelliteData> internal constructor(
    private val sharedData: GridData,
    override val cubeCoordinate: org.hexworks.mixite2.core.api.CubeCoordinate,
    private val hexagonDataStorage: HexagonDataStorage<T>) : Hexagon<T> {

    override val vertices: List<Double>
    override val points: List<Point>
    override val externalBoundingBox: Rectangle
    override val internalBoundingBox: Rectangle
    override val center: Point = calculateCenter()

    override val id: String
        get() = cubeCoordinate.toAxialKey()

    override val gridX: Int
        get() = cubeCoordinate.gridX

    override val gridY: Int
        get() = cubeCoordinate.gridY

    override val gridZ: Int
        get() = cubeCoordinate.gridZ

    override val satelliteData: Maybe<T>
        get() = hexagonDataStorage.getSatelliteDataBy(cubeCoordinate)

    init {
        this.points = calculatePoints()
        val x1 = points[3].coordinateX
        val y1 = points[2].coordinateY
        val x2 = points[0].coordinateX
        val y2 = points[5].coordinateY

        externalBoundingBox = Rectangle(x1, y1, x2 - x1, y2 - y1)
        internalBoundingBox = Rectangle((center.coordinateX - 1.25 * sharedData.radius / 2),
                (center.coordinateY - 1.25 * sharedData.radius / 2),
                (1.25f * sharedData.radius),
                (1.25f * sharedData.radius))

        this.vertices = ArrayList(12)
        for (point in points) {
            vertices.add(point.coordinateX)
            vertices.add(point.coordinateY)
        }
    }

    private fun calculateCenter(): Point {
        return if (HexagonOrientation.FLAT_TOP.equals(sharedData.orientation)) {
            Point.fromPosition(
                    cubeCoordinate.gridX * sharedData.hexagonWidth + sharedData.radius,
                    cubeCoordinate.gridZ * sharedData.hexagonHeight + cubeCoordinate.gridX * sharedData.hexagonHeight / 2 + sharedData.hexagonHeight / 2
            )
        } else {
            Point.fromPosition(
                    cubeCoordinate.gridX * sharedData.hexagonWidth + cubeCoordinate.gridZ * sharedData.hexagonWidth / 2 + sharedData.hexagonWidth / 2,
                    cubeCoordinate.gridZ * sharedData.hexagonHeight + sharedData.radius
            )
        }
    }

    private fun calculatePoints(): List<Point> {
        val points = ArrayList<Point>(6)
        for (i in 0..5) {
            val angle = 2 * PI / 6 * (i + sharedData.orientation.coordinateOffset)
            val x = center.coordinateX + sharedData.radius * cos(angle)
            val y = center.coordinateY + sharedData.radius * sin(angle)
            points.add(Point.fromPosition(x, y))
        }
        return points
    }

    override fun setSatelliteData(data: T) {
        this.hexagonDataStorage.addCoordinate(cubeCoordinate, data)
    }

    override fun clearSatelliteData() {
        this.hexagonDataStorage.clearDataFor(cubeCoordinate)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false

        other as HexagonImpl<*>

        if (cubeCoordinate != other.cubeCoordinate) return false

        return true
    }

    override fun hashCode(): Int {
        return cubeCoordinate.hashCode()
    }

}
